### ISSUE TITLE
Clarification on child feature workflow

### ISSUE THREAD

**Issue by natecox** [git-town/git-town#1178]
I'm trying to understand how parent/child branches work within git-town, but the documentation is a bit confusing.

There's a comment [here](https://github.com/Originate/git-town/blob/master/documentation/commands/ship.md) indicating that you can't ship a child:
> Only shipping of direct children of the main branch is allowed. To ship a nested child branch, all ancestor branches have to be shipped or killed.

So, what is the proposed workflow for child branches? Given a feature tree which looks like this:
```
Feature A
└── Feature B
    ├── Feature C
    └── Feature D

```
I would assume that 
- Feature `B` would be dependent on `C` and `D` for completion
- `C` and `D` would be merged into `B` as they were completed, and 
- `B` would be merged back into `A` when it was done.

Given that you can't `ship` B, C, or D into their parent branches, is there an official workflow from Town for this use-case, or should I just use standard git commands to issue merges? It's fairly easy to issue pull requests for the features and merge via github after code review, but I have read elsewhere that Town prefers you to _not_ use github's merge and instead use `ship`, so this seems a bit counter-intuitive and I'm assuming I'm just missing something obvous.

**Comment by charlierudolph** [comment]
Given `Feature A` is branched off your main development branch, Git Town is based around the idea that you ship A, then B, then C / D. 
```
Main
  Feature A
    Feature B
      Feature C
      Feature D
```
Shipping Feature A results in
```
Main
  Feature B
    Feature C
    Feature D
```
Shipping Feature B results in
```
Main
  Feature C
  Feature D
```
The nested branch feature was built with the idea that you want to continue working while you are getting a code review for Feature A and that A is a standalone feature. 

If you don't want to ship A, until B, C, D are merged into it then you would make A a perennial branch but still need to ship B into it and then C / D.

As the message says you are always shipping branches that are children of your main branch (or a perennial branch) 

---

Given all this, any suggestions on how to update the comment to make things more clear? 

Alternatively do you think we should remove the restriction for shipping and update it to always ship a child into its parent? The current restriction comes from the idea that shipping is only for delivering features which we believe is when you merge into the main branch (or a perennial branch)

**Comment by natecox** [comment]
**Author's note**: this turned out a bit longer than I intended. I'll keep the whole message below the divide, as I think it provides context, but for the purpose of brevity: 
Yes, I think an appropriate compromise here would be to remove shipping restrictions and always ship into the parent. I think it'd be awesome to keep the restrictions as described below, but maybe it's more red-tape than necessary.

----


I can see where you're coming from in the shipping order, but I don't think I would have come to that conclusion naturally. If, for example, I'm waiting on code review for `Feature A` before I start `Feature B`, I'd just `hack` a new feature off of my main branch instead of trying to `append` or `prepend`.

The way my team works, normally our features are independent of each other and don't need to be coupled; meaning we all just `hack` our own features off of the main branch. However, occasionally a very large feature is requested which may have multiple "parts" to it. Maybe a clearer representation is:
```
Epic Feature
    Subsystem A
        Subsystem A-1
        Subsystem A-2
    Subsystem B
```

Basically, `Epic Feature` can't be completed until subsystems `A` and `B` are complete, and subsequently, subsystem `A` can't be completed until `A-1` and `A-2` are complete. If we engineer the solution correctly, all five of the tree items above could be completed by a different team member and some or all of the work could be done in parallel. 

In this scenario, it would make sense to me that Subsystem `A-2` is a "child" of `A`, etc.; I think this is where the nomenclature is confusing for me. Also, in this scenario, it would make sense that shipping restrictions _are_ in place, just in the opposite order that is currently being used. E.g., `subsystem A` can't be merged into `epic feature` until `a-1` and `a-2` are both complete. Which we can read more simply as _features cannot be shipped while they have existing children"_

Long story short, I don't think that the current functionality is _wrong_, but I do believe that my scenario is valid and not mutually exclusive as well. 

Currently, we use github PRs to facilitate this, but it would be fantastic if town could support the workflow natively.

A compromise solution may be to keep the existing feature and naming, but simply remove the shipping restriction and define shipping as _into the parent branch_ rather than _into the main brach_ as you have suggested. I do see value in the restriction from both sides, but it may be that the restriction is more red-tape than it's worth when you consider the overhead of adding in a nearly-duplicate feature with different naming.

**Comment by charlierudolph** [comment]
@kevgo thoughts on this? I'm good opening up the workflow so children ship into their parent. If you want to keep the current workflow, I think we should make it a configuration option, so you can do this if desired.

---

> Edit: In retrospect, it occurs to me that git-town configs aren't shared between users. With this in mind, enforcing restrictions on ship as relates to team-play is going to be strained anyways; you would have to tell everyone else, "Oh, make sure you mark X as a [child|perrenial branch|etc] of y," which is a bit brittle. I think this reinforces to me that restrictions should just be removed... or that a standardized .git-town which can be committed to the repo is introduced—which I would fully support.

We originally had a checked in config for some pieces, but I wanted to remove it in favor of git configs. The branch hierarchy can't really be checked in in a clean way. You don't need the config once the branch is shipped but it would be nice that when you check it out it knows what the parent branch is. We could try adding something like it and tell people to treat it similar to lock/binary file that should never be reviewed. We could make it opt-in to try it out.

**Comment by natecox** [comment]
I would support a `git-town.lock` file as a part of the checked-in config. Aside from the workflow issues documented above, I think the only real frustration that we've had with town as a team is that when we pull over our peer's code for review and testing, every time you do that first `sync` you have to pick the parent branch, and occasionally it's not obvious. Granted, this is a minimal issue weighed against how much we've loved using town in general, but it would be a pretty cool quality-of-life improvement.

**Comment by kevgo** [comment]
TLDR: I'm fine with a `git ship --force` flag that overrides the safety checks. That's a good thing to have anyways imo.

Given that this "building a complex feature in small pieces that have to be all merged together" scenario is something that happens only occasionally for @natecox, and we haven't heard much demand from others for this, I'm hesitant to add dedicated support for it. I consider this approach more like "bending the best practice in particular edge cases". This approach has several negative side effects in my understanding:
- it results in a long-running feature branch: this is blocking other larger changes in unrelated branches, for example refactorings. Complex, long-running feature branches almost always end up as a bottleneck for the entire team.
- changes have to be reviewed several times: the change in 'Subsystem A-1' shows up again in the PR for `Subsystem A` as well as the PR for `Epic Feature`. The last PR mentioned here is probably so large that it is impossible to review in meaningful ways. Without discipline, changes made directly to the `Epic Feature` branch are falling through the review process here

We also sometimes run into the temptation to do this. Here are some strategies to avoid doing this:
- feature flags: build 

- acknowledge that this is a team-wide bottleneck and commit to pushing through building this feature in many small steps, with a broken main branch for a short duration. This means building and merging `Subsystem A-1`, then `Subsystem A-2`, then `Subsystem B`, then tidying up the remainder of the Epic feature and making everything work

- feature flags: build the

**Comment by natecox** [comment]
@kevgo I'm not sure I follow your arguments here.

If you accept that `Epic Feature` is the feature that needs building, then it's going to be in development for as long as it takes to build. Splitting that feature into smaller parts which can be worked on in parallel should _increase_ the speed of development for that feature, rather than decrease it. 

The alternative would be to use a single branch and have multiple people working on it together. In my opinion, sharing a code base between subsystems _compounds_ the downsides in the complexity of reviewing—which you've noted above—by forcing them to be evaluated only as a whole; rather than individually with a narrower focus and then together to ensure they've been pieced together correctly. The goal is to reduce the cognitive overhead required to assess each part accurately.

I also feel like you may be starting from a false clause, in that `Epic Feature` would block unrelated features from being shipped; unless you're suggesting that there should only ever be one feature branch open at a time. In large projects, the team may be working on three or four completely unrelated tasks at once; meaning that `Epic Feature` being a long-running development would in no way hinder any number of other features from being developed in parallel.

**Comment by charlierudolph** [comment]
I personally don't really like `--force` as other safety checks are still present. I'd rather have something like `-p, --to-parent` which means you want to ship to your parent branch instead of main/perennial. We could also have a configuration option where this is implied anytime you run `git ship`. How does that sound?

Also, I'll create a new issue to discuss a lockfile

**Comment by kevgo** [comment]
`-p` sounds good to me

**Comment by lud** [comment]
@natecox I can see "natecox closed this as completed"

What was the final decision on this? I am not sure I can find anything in the documentation.

Thank you.

**Comment by natecox** [comment]
@lud I don’t think there was a resolution here. I no longer use git-town (partly because I felt like it overly impeded my workflow as noted in the above thread) and closed this in a general cleanup of my open issues.

**Comment by lud** [comment]
Oh, I missed the fact that you were the author of the issue.

Thank you anyway :)

**Comment by lud-wj** [comment]
@kevgo today I am facing a big algorithmic problem and I want to tackle it. But I am not sure I go to the right direction. It seems to work, but I have another idea that I want to try, and I would like to compare. So being able to append, switch back to parent and append another, and finally ship a child to the parent with the `-p` option would be great.

Of course I can juste merge and delete the branch manually but when you're used to the power of `git ship` it's too bad :D

I have never touched a Goglang file so I'm not really ready to help but I would like this issue to reopen. Or create another one maybe?

**Comment by kevgo** [comment]
@lud you describe a convincing scenario where my suggestion of using feature flags and trunk-based development wouldn't work well. 

Since this ticket here is just a question and the author closed it, I have created a new ticket specifically for tracking the implementation of the feature to ship into non-perennial parent branches: https://github.com/git-town/git-town/issues/2605

I'm in the middle of a few large-scale cleanups before I'll get to building this. For the time being, you could emulate the behavior you want by shipping via the GitHub UI, i.e. clicking the "Merge" button on GitHub. Then run `git prune-branches` to remove the shipped branch from your machine. Once Git Town 10.0 gets released (hopefully next week), you won't need to run `git prune-branches` anymore.

How do you currently ship feature branches? Do you GitHub's web UI or do you run `git ship`?

**Comment by lud-wj** [comment]
On my current project I am using `git ship`. To compare branches I do not use github at all, I just compared on my machine and merged manually to the parent feature branch. We are not using pull requests yet, as it's a young project and we are only 2 working on it.

Glad to see this new feature is tracked. Good luck for v10!

**Comment by kevgo** [comment]
Thanks! FYI v10 is released now.
